test/test_tree.py

```diff
@@ -1125,6 +1126,120 @@ def test_max_levels_error(actx_factory):
+
+    from pytools.obj_array import make_obj_array
+    nodes = discr.nodes()
+    x = make_obj_array([flatten(coords, actx).with_queue(actx.queue)
+                        for coords in nodes])
+
+    return x, q
+
+
+@pytest.mark.parametrize("dim", [1, 2, 3])
+@pytest.mark.parametrize("order", [1, 2, 3])
+@pytest.mark.parametrize("nlevels", [1, 4])
+def test_uniform_tree_of_boxes(ctx_factory, dim, order, nlevels):
+    from boxtree.tree_build import make_tob_root, uniformly_refined
+    lower_bounds = np.random.rand(dim)
+    radius = np.random.rand() + 0.1
+    upper_bounds = lower_bounds + radius
+    tob = make_tob_root(dim=dim, bbox=[lower_bounds, upper_bounds])
+
+    for _ in range(nlevels - 1):
+        tob = uniformly_refined(tob)
+
+    from arraycontext import PyOpenCLArrayContext
+    queue = cl.CommandQueue(ctx_factory())
+    actx = PyOpenCLArrayContext(queue)
+
+    x, q = make_global_leaf_quadrature(actx, tob, order)
+
+    # integrates 1 exactly
+    assert np.isclose(sum(q.get()), radius**dim)
+
+
+@pytest.mark.parametrize("dim", [1, 2, 3])
+@pytest.mark.parametrize("order", [1, 2, 3, 4])
+def test_uniform_tree_of_boxes_convergence(ctx_factory, dim, order):
+    from boxtree.tree_build import make_tob_root, uniformly_refined
+    radius = np.pi
+    lower_bounds = np.zeros(dim) - radius/2
+    upper_bounds = lower_bounds + radius
+    tob = make_tob_root(dim=dim, bbox=[lower_bounds, upper_bounds])
+
+    min_level = 0
+    max_level = 1
+
+    for _ in range(min_level):
+        tob = uniformly_refined(tob)
+
+    # integrate cos(0.1*x + 0.2*y + 0.3*z + e) over [-pi/2, pi/2]**dim
+    qexact_table = {
+        1: 20 * np.sin(np.pi/20) * np.cos(np.e),
+        2: 50 * (np.sqrt(5) - 1) * np.sin(np.pi/20) * np.cos(np.e),
+        3: 250/3 * (np.sqrt(10 - 2*np.sqrt(5)) - 2) * np.cos(np.e)
+    }
+    qexact = qexact_table[dim]
+
+    from pytools.convergence import EOCRecorder
+    eoc_rec = EOCRecorder()
+
+    from arraycontext import PyOpenCLArrayContext
+    queue = cl.CommandQueue(ctx_factory())
+    actx = PyOpenCLArrayContext(queue)
+
+    for _ in range(min_level, max_level + 1):
+        x, q = make_global_leaf_quadrature(actx, tob, order)
+        x, q = (np.array([xx.get() for xx in x]), q.get())
+
+        inner = np.ones_like(q) * np.e
+        for iaxis in range(dim):
+            inner += (iaxis + 1) * 0.1 * x[iaxis]
+        f = np.cos(inner)
+        qh = np.sum(f * q)
+        err = abs(qexact - qh)
+
+        if err < 1e-14:
+            break  # eoc will be off after hitting machine epsilon
+
+        # under uniform refinement, last box is always leaf
+        eoc_rec.add_data_point(tob.get_box_size(-1), err)
+        tob = uniformly_refined(tob)
+
+    if len(eoc_rec.history) > 1:
+        # Gauss quadrature is exact up to degree 2q+1
+        eps = 0.01
+        assert eoc_rec.order_estimate() >= 2*order + 2 - eps
+    else:
+        print(err)
+        assert err < 1e-14
+
+# }}}
```

Comment:
Add a test that a traversal can be built off of (a CL version) of `TreeOfBoxes`.